jpyinterpreter/src/main/java/org/optaplanner/jpyinterpreter/StackMetadata.java

```diff
@@ -196,6 +196,12 @@ public StackMetadata push(ValueSourceInfo type) {
         return out;
     }
 
+    public StackMetadata set(int index, ValueSourceInfo type) {
+        StackMetadata out = copy();
+        out.stackValueSources.set(stackValueSources.size() - index - 1, type);
```

Comment:
Once the calculation reaches this level of complexity, I think that declaring a variable for it with a descriptive name makes the code easier to understand.